src/ppom-storage.ts

```diff
@@ -32,89 +34,137 @@ export type StorageKey = {
-
-  readonly #readMetadata: () => FileMetadataList;
-
-  readonly #writeMetadata: (metadata: FileMetadataList) => void;
-
-  /**
-   * Creates a PPOMStorage instance.
-   *
-   * @param options - The options passed to the function.
-   * @param options.storageBackend - The storage backend to use for the local storage.
-   * @param options.readMetadata - A function to read the metadata from the local storage.
-   * @param options.writeMetadata - A function to write the metadata to the local storage.
-   */
-  constructor({
-    storageBackend,
-    readMetadata,
-    writeMetadata,
-  }: {
-    storageBackend: StorageBackend;
-    readMetadata: () => FileMetadataList;
-    writeMetadata: (metadata: FileMetadataList) => void;
-  }) {
-    this.#storageBackend = storageBackend;
-    this.#readMetadata = readMetadata;
-    this.#writeMetadata = writeMetadata;
+const jsonStringToArrayBuffer = (json: string | undefined) => {
+  if (json) {
+    const dataToUintArray = new Uint8Array(
+      [...json].map((character) => character.charCodeAt(0)),
+    );
+    return dataToUintArray.buffer;
   }
 
-  /**
-   * Sync the metadata with the version info from the cdn.
-   * 1. Remove the files that are not readable (e.g. corrupted or deleted).
-   * 2. Remove the files that are not in the cdn anymore.
-   * 3. Remove the files that are not up to date in the cdn.
-   * 4. Remove the files that are not in the local storage from the metadata.
-   * 5. Delete the files that are not in the metadata from the local storage.
-   *
-   * @param versionInfo - Version information of metadata files.
-   */
-  async syncMetadata(versionInfo: FileMetadataList): Promise<FileMetadataList> {
-    const metadata = this.#readMetadata();
-    const syncedMetadata: FileMetadataList = [];
-
-    for (const fileMetadata of metadata) {
+  return undefined;
+};
+
+/**
+ * Read the file from the local storage.
+ * 1. Check if the file exists in the local storage.
+ * 2. Check if the file exists in the metadata.
+ *
+ * @param name - Name assigned to storage.
+ * @param chainId - ChainId for which file is queried.
+ * @param readState - Controller read state function.
+ */
+export const readFile = async (
+  name: string,
+  chainId: string,
+  readState: (key?: keyof PPOMState) => Partial<PPOMState>,
+): Promise<ArrayBuffer> => {
+  const { storageMetadata, fileStorage } = readState();
+
+  const fileMetadata = storageMetadata?.find(
+    (file: any) => file.name === name && file.chainId === chainId,
+  );
+  if (!fileMetadata) {
+    throw new Error(`File metadata (${name}, ${chainId}) not found`);
+  }
+
+  const stateFileStorage = fileStorage as Record<string, string>;
+
+  const data = jsonStringToArrayBuffer(stateFileStorage[`${name}_${chainId}`]);
+
+  await validateChecksum(
+    {
+      name,
+      chainId,
+    },
+    fileMetadata.checksum,
+    data,
+  );
+  if (!data) {
+    throw new Error(`Storage File (${name}, ${chainId}) not found`);
+  }
```

Comment:
This check can be before `validateChecksum`

Reply:
For consistency, doing the check inside validateChecksum as we also need to check this during writeFile